### PR TITLE
Azure devops client credentials

### PR DESCRIPTION
**Pull request by noydavidi**
## Contributing to Cortex XSOAR Content
Make sure to register your contribution by filling the [contribution registration form](https://forms.gle/XDfxU4E61ZwEESSMA)

**The Pull Request will be reviewed only after the contribution registration form is filled.**

## Status
- [x] In Progress
- [ ] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: [link to the issue](https://jira-dc.paloaltonetworks.com/browse/CIAC-10639)

## Description
Adding a client credentials flow for Azure DevOps integration

## Must have
- [ ] Tests
- [ ] Documentation

### CODE REVIEW

File: Packs/AzureDevOps/Integrations/AzureDevOps/AzureDevOps_description.md [hunk 1]
```diff
@@ -1,26 +1,39 @@
-### Authenticating
-To connect to the Azure DevOps:
+In order to connect to the Azure DevOps using either Cortex XSOAR Azure App or the Self-Deployed Azure App, use one of the following methods:
 
-1. In the Azure Portal, add a new Azure App Registration. To add the registration, refer to the following [Microsoft article](https://docs.microsoft.com/en-us/azure/active-directory/develop/quickstart-register-app).
-2. Add the following permissions to your registered app:
+1. *Authorization Code Flow* (Recommended).
+2. *Device Code Flow*.
+3. *Client Credentials Flow*.
```
Comment (ShirleyDenkberg) [inline Packs/AzureDevOps/Integrations/AzureDevOps/AzureDevOps_description.md:5]:
```suggestion
- *Authorization Code Flow* (Recommended).
- *Device Code Flow*.
- *Client Credentials Flow*.
```

File: Packs/AzureDevOps/Integrations/AzureDevOps/AzureDevOps_description.md [hunk 2]
```diff
@@ -31,14 +44,28 @@ For a Authorization Code configuration:
...
+   1. In the **Authentication Type** field, select the **Client Credentials** option.
+   2. In the **Application ID** field, enter your Client/Application ID.
+   3. In the **Tenant ID** field, enter your Tenant ID .
+   4. In the **Client Secret** field, enter your Client Secret.
+   5. Click **Test** to validate the URLs, token, and connection
+   6. Save the instance.
     
 ### Testing authentication and connectivity
-To test your authentication and connectivity to the Azure DevOps service run the ***!azure-devops-auth-test*** command.
+If you are using Device Code Flow or Authorization Code Flow, for testing your authentication and connectivity to the Azure DevOps service run the ***!azure-devops-auth-test*** command. If you are using Client Credentials Flow click on **Test** when you are configuring the instance.
```
Comment (ShirleyDenkberg) [inline Packs/AzureDevOps/Integrations/AzureDevOps/AzureDevOps_description.md:71]:
```suggestion
If you are using Device Code Flow or Authorization Code Flow, for testing your authentication and connectivity to the Azure DevOps service run the ***!azure-devops-auth-test*** command. 

If you are using Client Credentials Flow, click **Test** when you are configuring the instance.
```

File: Packs/AzureDevOps/Integrations/AzureDevOps/README.md [hunk 3]
```diff
@@ -21,12 +20,86 @@ Manage Git repositories in Azure DevOps Services. Integration capabilities inclu
...
+    | Application redirect URI | For user-auth mode or client credentials. | False |
     | Authorization code | for user-auth mode - received from the authorization step. see Detailed Instructions \(?\) section | False |
 4. Click **Test** to validate the URLs, token, and connection.
+
+
+In order to connect to the Azure DevOps using either Cortex XSOAR Azure App or the Self-Deployed Azure App, use one of the following methods:
+
+1. *Authorization Code Flow* (Recommended).
+2. *Device Code Flow*.
+3. *Client Credentials Flow*.
```
Comment (ShirleyDenkberg) [inline Packs/AzureDevOps/Integrations/AzureDevOps/README.md:35]:
```suggestion
- *Authorization Code Flow* (Recommended).
- *Device Code Flow*.
- *Client Credentials Flow*.
```

File: Packs/AzureDevOps/Integrations/AzureDevOps/README.md [hunk 4]
```diff
@@ -21,12 +20,86 @@ Manage Git repositories in Azure DevOps Services. Integration capabilities inclu
...
+To configure a Microsoft integration that uses this authorization flow with a self-deployed Azure application:
+   1. In the **Authentication Type** field, select the **Client Credentials** option.
+   2. In the **Application ID** field, enter your Client/Application ID.
+   3. In the **Tenant ID** field, enter your Tenant ID .
+   4. In the **Client Secret** field, enter your Client Secret.
+   5. Click **Test** to validate the URLs, token, and connection
+   6. Save the instance.
+    
+### Testing authentication and connectivity
+If you are using Device Code Flow or Authorization Code Flow, for testing your authentication and connectivity to the Azure DevOps service run the ***!azure-devops-auth-test*** command. If you are using Client Credentials Flow click on **Test** when you are configuring the instance.
```
Comment (ShirleyDenkberg) [inline Packs/AzureDevOps/Integrations/AzureDevOps/README.md:101]:
```suggestion
If you are using Device Code Flow or Authorization Code Flow, for testing your authentication and connectivity to the Azure DevOps service run the ***!azure-devops-auth-test*** command. 

If you are using Client Credentials Flow, click **Test** when you are configuring the instance.
```

**Comment by ShirleyDenkberg** [comment]
@JudahSchwartz Doc review completed.

File: Packs/AzureDevOps/Integrations/AzureDevOps/AzureDevOps.py [hunk 5]
```diff
@@ -67,12 +72,13 @@ def __init__(self, client_id: str, organization: str, verify: bool, proxy: bool,
...
             auth_id=client_id,
-            token_retrieval_url='https://login.microsoftonline.com/organizations/oauth2/v2.0/token',
+            token_retrieval_url='https://login.microsoftonline.com/organizations/oauth2/v2.0/token' if 'Device Code' in
+                                                                                                       auth_type else None,
             grant_type=GRANT_BY_CONNECTION[auth_type],
             base_url=f'https://dev.azure.com/{organization}',
             verify=verify,
             proxy=proxy,
-            scope=AZURE_DEVOPS_SCOPE,
+            scope=SCOPE_CLIENT_CREDENTIALS_FLOW if 'Client Credentials' in auth_type else SCOPE_DEVICE_AUTH_FLOW,
```
Comment (michal-dagan) [inline Packs/AzureDevOps/Integrations/AzureDevOps/AzureDevOps.py:81]: I think the new scope is also relevant for 'Authorization Code' flow, so maybe change it to if not ''Device Code' (we can check it)

**Review by michal-dagan — APPROVED** [review]
Nice!

**Review by JudahSchwartz — APPROVED** [review]

**Pull request merged**